### DISCUSSION TITLE
Modal windows for desktop apps

### DISCUSSION THREAD

**Discussion by mikeparker104** [CommunityToolkit/Maui#840] [New Feature Discussions]
Ability to display a traditional desktop modal window in a .NET MAUI app that:

- Can be moved beyond the confines of a specific window e.g., so it can be displayed next to the window or on another monitor
- Prevents interaction with all other application windows until that modal window has been closed
- Keeps focus (and appears on top of the other application windows) even when other application windows receive input
- Does not appear alongside other windows in the taskbar and/or switcher

Existing options, within .NET MAUI and other third-party components, only support the display of a Page and/or dialog that is constrained to the Window it was opened from.

The expectation is that this requires use of the following platform-specific APIs.

**MacCatalyst**

A modal event loop can be started and stopped for a specific [NSWindow](https://developer.apple.com/documentation/appkit/nswindow?language=objc) using the following APIs from the [NSApplication](https://developer.apple.com/documentation/appkit/nsapplication?language=objc) class:

- [runModal(for:)](https://developer.apple.com/documentation/appkit/nsapplication/1428436-runmodal)
- [stopModal](https://developer.apple.com/documentation/appkit/nsapplication/1428489-stopmodal)

While the [NSApplication](https://developer.apple.com/documentation/appkit/nsapplication?language=objc) and [NSWindow](https://developer.apple.com/documentation/appkit/nswindow) types cannot be used directly in [MacCatalyst](https://developer.apple.com/documentation/uikit/mac_catalyst?language=objc), it's possible to get a reference to those types and invoke the requisite functionality on them via selectors.

**WinUI3**

There's an [open proposal](https://github.com/microsoft/microsoft-ui-xaml/issues/885) for adding modal dialog support to [WinUI3](https://docs.microsoft.com/windows/apps/winui/winui3/) but it does not currently support this concept directly. However, the requirements might be achieved through a combination of [Win32](https://docs.microsoft.com/windows/win32/apiindex/windows-api-list) and [Windows App SDK Interop](https://learn.microsoft.com/windows/apps/desktop/modernize/winrt-com-interop-csharp#available-as-part-of-the-windows-app-sdk) APIs alongside the [Windows App SDK](https://learn.microsoft.com/windows/windows-app-sdk) components as described in an answer to a [recent forum question](https://docs.microsoft.com/answers/questions/910658/winui3how-to-show-a-model-window.html). Notable members, types, and APIs include:

***Types and Members***
- [AppWindow.IsShownInSwitchers](https://learn.microsoft.com/windows/windows-app-sdk/api/winrt/microsoft.ui.windowing.appwindow.isshowninswitchers?view=windows-app-sdk-1.1)
- [AppWindow.Presenter](https://learn.microsoft.com/windows/windows-app-sdk/api/winrt/microsoft.ui.windowing.appwindow.presenter?view=windows-app-sdk-1.1)
- [OverlappedPresenter.IsMaximizable](https://learn.microsoft.com/windows/windows-app-sdk/api/winrt/microsoft.ui.windowing.overlappedpresenter.ismaximizable?view=windows-app-sdk-1.1)
- [OverlappedPresenter.IsMinimizable](https://learn.microsoft.com/windows/windows-app-sdk/api/winrt/microsoft.ui.windowing.overlappedpresenter.isminimizable?view=windows-app-sdk-1.1)
- [OverlappedPresenter.IsResizable](https://learn.microsoft.com/windows/windows-app-sdk/api/winrt/microsoft.ui.windowing.overlappedpresenter.isresizable?view=windows-app-sdk-1.1)

***WIn32 APIs and Constants***
- [GetWindow](https://learn.microsoft.com/windows/win32/api/winuser/nf-winuser-getwindow)
- [EnableWindow](https://learn.microsoft.com/windows/win32/api/winuser/nf-winuser-enablewindow)
- [SetWindowLong](https://learn.microsoft.com/windows/win32/api/winuser/nf-winuser-setwindowlonga)
- [WindowStyles](https://learn.microsoft.com/windows/win32/winmsg/window-styles)

Would others consider this a valuable addition to the Community Toolkit?

**Comment by mikeparker104** [comment]
Created initial feature proposal here: https://github.com/CommunityToolkit/Maui/issues/842

**Comment by VladislavAntonyuk** [comment]
It can be achieved by Popup. Why do you need a new window?

**Reply by mikeparker104** [reply]
Hi @VladislavAntonyuk, thanks for your reply. To confirm my understanding, you're saying that the Popup can already be displayed outside of the window it was opened from/within? Or are you asking why modal windows would be required for multi-window desktop apps in general?

**Reply by VladislavAntonyuk** [reply]
I am just asking how you plan to use it?
What is your scenario that you cannot achieve with Popup?

**Reply by mikeparker104** [reply]
In multi-window desktop apps, a modal window is often used to ensure a task or flow is completed before further interaction with application functionality is able to take place on any window in the application. For example, in Visual Studio when a user needs to sign in, switch accounts, or change some global settings.

**Reply by pictos** [reply]
@mikeparker104 can't you archive this with the popup control?

**Reply by mikeparker104** [reply]
@pictos My understanding is the popup doesn't prevent interaction with other open application windows and also can't be moved beyond the confines of a specific window e.g., so it can be displayed next to the window or on another monitor.

**Comment by faustmet** [comment]
A modal may be required when we want to ensure that this flow is completed before another interaction, especially where we need a confirmation before any future action is taken. Also, we would like to prevent background or offscreen popups from collecting in general.

**Comment by mikeparker104** [comment]
Thanks for discussing during the [February 2023 .NET MAUI Community Toolkit Monthly Standup](https://www.youtube.com/watch?v=jzk8uqBB9zo&t=1044s). Some further input on potential handling on non-desktop platforms.

Looking at [features for building great desktop apps](https://devblogs.microsoft.com/dotnet/5-dotnet-maui-desktop-features/#multi-window), and how those are handled in MAUI, the APIs and/or types are exposed to all TFMs but are ignored (or fail silently) on platforms where they're not suitable/supported. On this basis, developers would either do nothing or handle the flow/UI differently based on whether they're targeting mobile or desktop. However, in this case, the current [INavigation.PushModalAsync](https://learn.microsoft.com/dotnet/api/microsoft.maui.controls.inavigation.pushmodalasync?view=net-maui-7.0) behavior feels like a suitable fallback for those platforms that don't support multi-window (and modal windows). 

As a starting point, one approach might be to extend [INavigation](https://learn.microsoft.com/dotnet/api/microsoft.maui.controls.inavigation?view=net-maui-7.0) with a version of  [INavigation.PushModalAsync](https://learn.microsoft.com/dotnet/api/microsoft.maui.controls.inavigation.pushmodalasync?view=net-maui-7.0) that will show the specified [Page](https://learn.microsoft.comdotnet/api/microsoft.maui.controls.page?view=net-maui-7.0) in a modal Window if supported but otherwise use the current method to display it within the current Window. 

### Indicative implementation

```csharp
namespace Microsoft.Maui.Controls;

public static class INavigationExtensions
{
    public static Task PushModalAsyncEx(this INavigation navigation, Page page)
    {
#if MACCATALYST || WINDOWS
        Application.Current.OpenWindow(new ModalWindow(page));
        return Task.CompletedTask;
#else
        return navigation.PushModalAsync(page);
#endif
    }
}
```

### Indicative usage

```csharp
public partial class MainPage : ContentPage
{
    void OnButtonClicked(object sender, EventArgs e)
        => Navigation.PushModalAsyncEx(new MyPage());
}
```

### Further considerations

Assuming demand continues to grow within the community for desktop focused features, one approach to consider in future might be splitting those capabilities out into a standalone version of the toolkit specific to desktop. For example, CommunityToolkit.Maui.Desktop.

Thoughts?

**Reply by mouralabank** [reply]
@mikeparker104  I totally agree with you, MAUI has a lot of potential for multi platform and with that the need for desktop-focused demands eventually, such as Modal, MDI, toolbar/ribbon, mouse events, keyboard events, options to configure a window (no titlebar, always on top, without minimizing, without maximizing, just maximized...), tabbed pages, control the mouse for accessibility (no clickable areas change the mouse to indicate this,  for example), custom dialogs

**Reply by danielancines** [reply]
This will be great for desktop applications, even WPF been an option, the implementation with Maui is great for mobile and it'll be amazing to hit all targets (windows, macos, linux (future maybe), iOs, Android and Web) with one code base. Maui is great and this option will put Maui on the top of options. Thrilled to see what's next.

**Comment by CliffAgius** [comment]
Closed as answered.
